File: native/cocos/scene/SubModel.cpp
/****************************************************************************
 Copyright (c) 2021-2023 Xiamen Yaji Software Co., Ltd.

 http://www.cocos.com

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
****************************************************************************/

#include "scene/SubModel.h"
#include "core/Root.h"
#include "core/platform/Debug.h"
#include "pipeline/Define.h"
#include "pipeline/InstancedBuffer.h"
#include "renderer/pipeline/PipelineSceneData.h"
#include "renderer/pipeline/custom/RenderInterfaceTypes.h"
#include "renderer/pipeline/forward/ForwardPipeline.h"
#include "scene/Model.h"
#include "scene/Pass.h"
#include "scene/Shadow.h"

namespace cc {
namespace scene {

const ccstd::string INST_MAT_WORLD = "a_matWorld0";
const ccstd::string INST_SH = "a_sh_linear_const_r";

cc::TypedArray getTypedArrayConstructor(const cc::gfx::FormatInfo &info, cc::ArrayBuffer *buffer, uint32_t byteOffset, uint32_t length) {
    const uint32_t stride = info.size / info.count;
    switch (info.type) {
        case cc::gfx::FormatType::UNORM:
        case cc::gfx::FormatType::UINT: {
            switch (stride) {
                case 1: return cc::Uint8Array(buffer, byteOffset, length);
                case 2: return cc::Uint16Array(buffer, byteOffset, length);
                case 4: return cc::Uint32Array(buffer, byteOffset, length);
                default:
                    break;
            }
            break;
        }
        case cc::gfx::FormatType::SNORM:
        case cc::gfx::FormatType::INT: {
            switch (stride) {
                case 1: return cc::Int8Array(buffer, byteOffset, length);
                case 2: return cc::Int16Array(buffer, byteOffset, length);
                case 4: return cc::Int32Array(buffer, byteOffset, length);
                default:
                    break;
            }
            break;
        }
        case cc::gfx::FormatType::FLOAT: {
            return cc::Float32Array(buffer, byteOffset, length);
        }
        default:
            break;
    }
    return cc::Float32Array(buffer, byteOffset, length);
}

SubModel::SubModel() {
    _id = generateId();
}

const static uint32_t MAX_PASS_COUNT = 8;

void SubModel::update() {
    const auto &passes = *_passes;
    for (Pass *pass : passes) {
        pass->update();
    }
    _descriptorSet->update();

    if (_worldBoundDescriptorSet) {
        _worldBoundDescriptorSet->update();
    }
}

void SubModel::setPasses(const SharedPassArray &pPasses) {
    if (!pPasses || pPasses->size() > MAX_PASS_COUNT) {
        debug::errorID(12004, MAX_PASS_COUNT);
        return;
    }

    _passes = pPasses;
    flushPassInfo();

    const auto &passes = *_passes;
    // DS layout might change too
    if (_descriptorSet) {
        _descriptorSet->destroy();
        gfx::DescriptorSetInfo dsInfo;
        dsInfo.layout = passes[0]->getLocalSetLayout();
        _descriptorSet = _device->createDescriptorSet(dsInfo);
    }
}

gfx::Shader *SubModel::getShader(uint32_t index) const {
    if (index >= _shaders.size()) {
        return nullptr;
    }

    return _shaders[index];
}

Pass *SubModel::getPass(uint32_t index) const {
    auto &passes = *_passes;
    if (index >= passes.size()) {
        return nullptr;
    }

    return passes[index];
}

void SubModel::initialize(RenderingSubMesh *subMesh, const SharedPassArray &pPasses, const ccstd::vector<IMacroPatch> &patches) {
    _device = Root::getInstance()->getDevice();
    CC_ASSERT(!pPasses->empty());
    gfx::DescriptorSetInfo dsInfo;
    dsInfo.layout = (*pPasses)[0]->getLocalSetLayout();
    _inputAssembler = _device->createInputAssembler(subMesh->getIaInfo());
    _descriptorSet = _device->createDescriptorSet(dsInfo);

    const auto *pipeline = Root::getInstance()->getPipeline();
    const auto *occlusionPass = pipeline->getPipelineSceneData()->getOcclusionQueryPass();
    if (occlusionPass) {
        cc::gfx::DescriptorSetInfo occlusionDSInfo;
        occlusionDSInfo.layout = occlusionPass->getLocalSetLayout();
        _worldBoundDescriptorSet = _device->createDescriptorSet(occlusionDSInfo);
    }

    _subMesh = subMesh;
    ccstd::vector<IMacroPatch> tmp = patches;
    std::sort(tmp.begin(), tmp.end(), IMacroPatch::compare);
    _patches = tmp;
    _passes = pPasses;

    flushPassInfo();

    const auto &passes = *_passes;
    _priority = pipeline::RenderPriority::DEFAULT;

    // initialize resources for reflection material
    if (passes[0]->getPhase() == pipeline::getPhaseID("reflection")) {
        const auto *mainWindow = Root::getInstance()->getMainWindow();
        uint32_t texWidth = mainWindow->getWidth();
        uint32_t texHeight = mainWindow->getHeight();
        const uint32_t minSize = 512;
        if (texHeight < texWidth) {
            texWidth = minSize * texWidth / texHeight;
            texHeight = minSize;
        } else {
            texWidth = minSize;
            texHeight = minSize * texHeight / texWidth;
        }
        _reflectionTex = _device->createTexture(gfx::TextureInfo{
            gfx::TextureType::TEX2D,
            gfx::TextureUsageBit::STORAGE | gfx::TextureUsageBit::TRANSFER_SRC | gfx::TextureUsageBit::SAMPLED,
            gfx::Format::RGBA8,
            texWidth,
            texHeight,
        });
        _descriptorSet->bindTexture(pipeline::REFLECTIONTEXTURE::BINDING, _reflectionTex);

        const gfx::SamplerInfo samplerInfo{
            gfx::Filter::LINEAR,
            gfx::Filter::LINEAR,
            gfx::Filter::NONE,
            gfx::Address::CLAMP,
            gfx::Address::CLAMP,
            gfx::Address::CLAMP,
        };
        _reflectionSampler = _device->getSampler(samplerInfo);
        _descriptorSet->bindSampler(pipeline::REFLECTIONTEXTURE::BINDING, _reflectionSampler);
        _descriptorSet->bindTexture(pipeline::REFLECTIONSTORAGE::BINDING, _reflectionTex);
    }
}

void SubModel::destroy() {
    CC_SAFE_DESTROY_NULL(_descriptorSet);
    CC_SAFE_DESTROY_NULL(_inputAssembler);
    CC_SAFE_DESTROY_NULL(_worldBoundDescriptorSet);

    _priority = pipeline::RenderPriority::DEFAULT;

    _patches.clear();
    _globalPatches.clear();
    _subMesh = nullptr;
    _passes.reset();
    _shaders.clear();

    CC_SAFE_DESTROY_NULL(_reflectionTex);
    _reflectionSampler = nullptr;
}

void SubModel::onPipelineStateChanged() {
    const auto *pipeline = Root::getInstance()->getPipeline();
    ccstd::vector<IMacroPatch> pipelinePatches(pipeline->getMacros().begin(), pipeline->getMacros().end());
    ccstd::vector<IMacroPatch> globalPatches(_globalPatches.begin(), _globalPatches.end());
    if (pipelinePatches.empty() && globalPatches.empty()) {
        return;
    }

    std::sort(pipelinePatches.begin(), pipelinePatches.end(), IMacroPatch::compare);
    std::sort(globalPatches.begin(), globalPatches.end(), IMacroPatch::compare);
    if (std::equal(std::begin(pipelinePatches), std::end(pipelinePatches), std::begin(globalPatches), std::end(globalPatches))) {
        return;
    }
    _globalPatches = pipeline->getMacros();

    const auto &passes = *_passes;
    if (passes.empty()) return;

    for (Pass *pass : passes) {
        pass->beginChangeStatesSilently();
        pass->tryCompile(); // force update shaders
        pass->endChangeStatesSilently();
    }
    flushPassInfo();
}

void SubModel::onMacroPatchesStateChanged(const ccstd::vector<IMacroPatch> &patches) {
    if (patches.empty() && _patches.empty()) {
        return;
    }

    ccstd::vector<IMacroPatch> tmp = patches;
    std::sort(tmp.begin(), tmp.end(), IMacroPatch::compare);
    if (std::equal(std::begin(tmp), std::end(tmp), std::begin(_patches), std::end(_patches))) {
        return;
    }
    _patches = tmp;
    const auto &passes = *_passes;
    if (passes.empty()) return;
    for (Pass *pass : passes) {
        pass->beginChangeStatesSilently();
        pass->tryCompile(); // force update shaders
        pass->endChangeStatesSilently();
    }
    flushPassInfo();
}

void SubModel::onGeometryChanged() {
    if (!_subMesh) {
        return;
    }

    // update draw info
    const auto &drawInfo = _subMesh->getDrawInfo();
    if (drawInfo.has_value()) {
        _inputAssembler->setDrawInfo(drawInfo.value());
    }
}

void SubModel::updateInstancedAttributes(const ccstd::vector<gfx::Attribute> &attributes) {
    auto *pass = getPass(0);
    _instancedWorldMatrixIndex = -1;
    _instancedSHIndex = -1;
    if (!pass->getDevice()->hasFeature(gfx::Feature::INSTANCED_ARRAYS)) return;
    // free old data

    uint32_t size = 0;
    for (const gfx::Attribute &attribute : attributes) {
        if (!attribute.isInstanced) continue;
        size += gfx::GFX_FORMAT_INFOS[static_cast<uint32_t>(attribute.format)].size;
    }
    auto &attrs = _instancedAttributeBlock;
    attrs.buffer = Uint8Array(size);
    attrs.views.clear();
    attrs.attributes.clear();
    attrs.views.reserve(attributes.size());
    attrs.attributes.reserve(attributes.size());

    uint32_t offset = 0;

    for (const gfx::Attribute &attribute : attributes) {
        if (!attribute.isInstanced) continue;
        gfx::Attribute attr;
        attr.format = attribute.format;
        attr.name = attribute.name;
        attr.isNormalized = attribute.isNormalized;
        attr.location = attribute.location;
        attrs.attributes.emplace_back(attr);
        const auto &info = gfx::GFX_FORMAT_INFOS[static_cast<uint32_t>(attribute.format)];
        auto *buffer = attrs.buffer.buffer();
        auto typeViewArray = getTypedArrayConstructor(info, buffer, offset, info.count);
        attrs.views.emplace_back(typeViewArray);
        offset += info.size;
    }
    if (pass->getBatchingScheme() == BatchingSchemes::INSTANCING) {
        pass->getInstancedBuffer()->destroy();
    }
    _instancedWorldMatrixIndex = getInstancedAttributeIndex(INST_MAT_WORLD);
    _instancedSHIndex = getInstancedAttributeIndex(INST_SH);
}

void SubModel::updateInstancedWorldMatrix(const Mat4 &mat, int32_t idx) {
    auto &attrs = _instancedAttributeBlock.views;
    auto &v1 = ccstd::get<Float32Array>(attrs[idx]);
    auto &v2 = ccstd::get<Float32Array>(attrs[idx + 1]);
    auto &v3 = ccstd::get<Float32Array>(attrs[idx + +2]);
    const uint32_t copyBytes = sizeof(float) * 3;
    auto *buffer = const_cast<uint8_t *>(v1.buffer()->getData());

    uint8_t *dst = buffer + v1.byteOffset();
    memcpy(dst, mat.m, copyBytes);
    v1[3] = mat.m[12];

    dst = buffer + v2.byteOffset();
    memcpy(dst, mat.m + 4, copyBytes);
    v2[3] = mat.m[13];

    dst = buffer + v3.byteOffset();
    memcpy(dst, mat.m + 8, copyBytes);
    v3[3] = mat.m[14];
}

void SubModel::updateInstancedSH(const Float32Array &data, int32_t idx) {
    auto &attrs = _instancedAttributeBlock.views;
    const auto count = (pipeline::UBOSH::SH_QUADRATIC_R_OFFSET - pipeline::UBOSH::SH_LINEAR_CONST_R_OFFSET) / 4;
    auto offset = 0;

    for (auto i = idx; i < idx + count; i++) {
        auto &attr = ccstd::get<Float32Array>(attrs[i]);
        for (auto k = 0; k < 4; k++) {
            attr[k] = data[offset++];
        }
    }
}

void SubModel::flushPassInfo() {
    const auto &passes = *_passes;
    if (passes.empty()) return;
    if (!_shaders.empty()) {
        _shaders.clear();
    }
    _shaders.resize(passes.size());
    for (size_t i = 0; i < passes.size(); ++i) {
        _shaders[i] = passes[i]->getShaderVariant(_patches);
    }
}

void SubModel::setSubMesh(RenderingSubMesh *subMesh) {
    const auto &passes = *_passes;
    _inputAssembler->destroy();
    _inputAssembler->initialize(subMesh->getIaInfo());
    _subMesh = subMesh;
}

void SubModel::setInstancedAttribute(const ccstd::string &name, const float *value, uint32_t byteLength) {
    const auto &attributes = _instancedAttributeBlock.attributes;
    auto &views = _instancedAttributeBlock.views;
    for (size_t i = 0, len = attributes.size(); i < len; ++i) {
        const auto &attribute = attributes[i];
        if (attribute.name == name) {
            const auto &info = gfx::GFX_FORMAT_INFOS[static_cast<uint32_t>(attribute.format)];
            switch (info.type) {
                case gfx::FormatType::NONE:
                case gfx::FormatType::UNORM:
                case gfx::FormatType::SNORM:
                case gfx::FormatType::UINT:
                case gfx::FormatType::INT: {
                    CC_ABORT();
                } break;
                case gfx::FormatType::FLOAT:
                case gfx::FormatType::UFLOAT: {
                    CC_ASSERT(ccstd::holds_alternative<Float32Array>(views[i]));
                    auto &view = ccstd::get<Float32Array>(views[i]);
                    auto *dstData = reinterpret_cast<float *>(view.buffer()->getData() + view.byteOffset());
                    CC_ASSERT(byteLength <= view.byteLength());
                    memcpy(dstData, value, byteLength);
                } break;
                default:
                    break;
            }
        }
    }
}

int32_t SubModel::getInstancedAttributeIndex(const ccstd::string &name) const {
    const auto &attributes = _instancedAttributeBlock.attributes;
    for (index_t i = 0; i < static_cast<index_t>(attributes.size()); ++i) {
        if (attributes[i].name == name) {
            return i;
        }
    }
    return CC_INVALID_INDEX;
}

} // namespace scene
} // namespace cc

File: cocos/render-scene/scene/submodel.ts
/*
 Copyright (c) 2020-2023 Xiamen Yaji Software Co., Ltd.

 https://www.cocos.com/

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
*/

import { RenderingSubMesh } from '../../asset/assets/rendering-sub-mesh';
import { RenderPriority, UNIFORM_REFLECTION_TEXTURE_BINDING, UNIFORM_REFLECTION_STORAGE_BINDING,
    INST_MAT_WORLD, INST_SH, UBOSH, isEnableEffect } from '../../rendering/define';
import { BatchingSchemes, IMacroPatch, Pass } from '../core/pass';
import { DescriptorSet, DescriptorSetInfo, Device, InputAssembler, Texture, TextureType, TextureUsageBit, TextureInfo,
    Format, Sampler, Filter, Address, Shader, SamplerInfo, deviceManager,
    Attribute, Feature, FormatInfos, getTypedArrayConstructor } from '../../gfx';
import { errorID, Mat4, cclegacy } from '../../core';
import { getPhaseID } from '../../rendering/pass-phase';
import { Root } from '../../root';

const _dsInfo = new DescriptorSetInfo(null!);
const MAX_PASS_COUNT = 8;

export interface IInstancedAttributeBlock {
    buffer: Uint8Array;
    views: TypedArray[];
    attributes: Attribute[];
}

/**
 * @en A sub part of the model, it describes how to render a specific sub mesh.
 * It contains geometry information in [[RenderingSubMesh]] and all sort of rendering configuration like shaders, macro patches, passes etc.
 * @zh 组成模型对象的子模型，它用来描述如何渲染模型的一个子网格。
 * 它包含 [[RenderingSubMesh]] 代表的几何网格信息和所有渲染需要的数据，比如着色器程序，着色器宏定义，渲染 pass，等。
 */
export class SubModel {
    protected _device: Device | null = null;
    protected _passes: Pass[] | null = null;
    protected _shaders: Shader[] | null = null;
    protected _subMesh: RenderingSubMesh | null = null;
    protected _patches: IMacroPatch[] | null = null;
    protected _priority: RenderPriority = RenderPriority.DEFAULT;
    protected _inputAssembler: InputAssembler | null = null;
    protected _descriptorSet: DescriptorSet | null = null;
    protected _worldBoundDescriptorSet: DescriptorSet | null = null;
    protected _reflectionTex: Texture | null = null;
    protected _reflectionSampler: Sampler | null = null;
    protected _instancedAttributeBlock: IInstancedAttributeBlock = { buffer: null!, views: [], attributes: [] };
    protected _instancedWorldMatrixIndex = -1;
    protected _instancedSHIndex = -1;
    protected _useReflectionProbeType = 0;

    /**
     * @en
     * sub model's passes
     * @zh
     * 子模型的 passes
     * @param passes @en The passes @zh 设置的 passes
     */
    /**
     * @en Render passes for the sub-model
     * @zh 子模型的渲染 pass
     */
    set passes (passes) {
        const passLengh = passes.length;
        if (passLengh > MAX_PASS_COUNT) {
            errorID(12004, MAX_PASS_COUNT);
            return;
        }
        this._passes = passes;
        this._flushPassInfo();

        // DS layout might change too
        if (this._descriptorSet) {
            this._descriptorSet.destroy();
            _dsInfo.layout = passes[0].localSetLayout;
            this._descriptorSet = this._device!.createDescriptorSet(_dsInfo);
        }
    }

    get passes (): Pass[] {
        return this._passes!;
    }

    /**
     * @en Shaders for the sub-model, each shader corresponds to one of the [[passes]]
     * @zh 子模型的着色器程序列表，每个着色器程序对应其中一个渲染 [[passes]]
     */
    get shaders (): Shader[] {
        return this._shaders!;
    }

    /**
     * @en The rendering sub mesh for the sub-model, each sub-model can only have one sub mesh.
     * @zh 用于渲染的子网格对象，每个子模型只能包含一个子网格。
     */
    set subMesh (subMesh) {
        this._inputAssembler!.destroy();
        this._inputAssembler = this._device!.createInputAssembler(subMesh.iaInfo);
        this._subMesh = subMesh;
    }

    get subMesh (): RenderingSubMesh {
        return this._subMesh!;
    }

    /**
     * @en The rendering priority of the sub-model
     * @zh 子模型的渲染优先级
     */
    set priority (val) {
        this._priority = val;
    }

    get priority (): RenderPriority {
        return this._priority;
    }

    /**
     * @en The low level input assembler which contains geometry data
     * @zh 底层渲染用的输入汇集器，包含几何信息
     */
    get inputAssembler (): InputAssembler {
        return this._inputAssembler!;
    }

    /**
     * @en The descriptor set used for sub-model rendering
     * @zh 底层渲染子模型用的描述符集组
     */
    get descriptorSet (): DescriptorSet {
        return this._descriptorSet!;
    }

    /**
     * @en The descriptor set for world bound
     * @zh 用于存储世界包围盒的描述符集组
     */
    get worldBoundDescriptorSet (): DescriptorSet | null {
        return this._worldBoundDescriptorSet;
    }

    /**
     * @en The macro patches for the shaders
     * @zh 着色器程序所用的宏定义组合
     */
    get patches (): Readonly<IMacroPatch[] | null> {
        return this._patches;
    }

    /**
     * @en The instance attribute block, access by sub model
     * @zh 硬件实例化属性，通过子模型访问
     */
    get instancedAttributeBlock (): IInstancedAttributeBlock {
        return this._instancedAttributeBlock;
    }

    /**
     * @en Get or set instance matrix id, access by sub model
     * @zh 获取或者设置硬件实例化中的矩阵索引，通过子模型访问
     */
    set instancedWorldMatrixIndex (val: number) {
        this._instancedWorldMatrixIndex = val;
    }
    get instancedWorldMatrixIndex (): number {
        return this._instancedWorldMatrixIndex;
    }

    /**
     * @en Get or set instance SH id, access by sub model
     * @zh 获取或者设置硬件实例化中的球谐索引，通过子模型访问
     */
    set instancedSHIndex (val: number) {
        this._instancedSHIndex = val;
    }
    get instancedSHIndex (): number {
        return this._instancedSHIndex;
    }

    /**
     * @en Gets or sets the type of reflection probe, Used to process instance
     * @zh 获取或设置使用反射探针的类型，用于处理instance
     */
    set useReflectionProbeType (val) {
        this._useReflectionProbeType = val;
    }
    get useReflectionProbeType (): number {
        return this._useReflectionProbeType;
    }

    /**
     * @en
     * init sub model
     * @zh
     * 子模型初始化
     * @param subMesh @en The sub mesh @zh 子网格资源
     * @param passes @en The passes @zh 渲染的 passes
     * @param patches @en The shader's macro @zh 着色器的宏定义
     */
    public initialize (subMesh: RenderingSubMesh, passes: Pass[], patches: IMacroPatch[] | null = null): void {
        const root = cclegacy.director.root as Root;
        this._device = deviceManager.gfxDevice;
        _dsInfo.layout = passes[0].localSetLayout;

        this._inputAssembler = this._device.createInputAssembler(subMesh.iaInfo);
        this._descriptorSet = this._device.createDescriptorSet(_dsInfo);

        const pipeline = (cclegacy.director.root as Root).pipeline;
        const occlusionPass = pipeline.pipelineSceneData.getOcclusionQueryPass();
        if (occlusionPass) {
            const occlusionDSInfo = new DescriptorSetInfo(null!);
            occlusionDSInfo.layout = occlusionPass.localSetLayout;
            this._worldBoundDescriptorSet = this._device.createDescriptorSet(occlusionDSInfo);
        }

        this._subMesh = subMesh;
        this._patches = patches ? patches.sort() : null;
        this._passes = passes;

        this._flushPassInfo();

        this.priority = RenderPriority.DEFAULT;
        const r = cclegacy.rendering;
        // initialize resources for reflection material
        if (((!r || !r.enableEffectImport) && passes[0].phase === getPhaseID('reflection'))
        || (isEnableEffect() && passes[0].phaseID === r.getPhaseID(r.getPassID('default'), 'reflection'))) {
            let texWidth = root.mainWindow!.width;
            let texHeight = root.mainWindow!.height;
            const minSize = 512;

            if (texHeight < texWidth) {
                texWidth = minSize * texWidth / texHeight;
                texHeight = minSize;
            } else {
                texWidth = minSize;
                texHeight = minSize * texHeight / texWidth;
            }

            this._reflectionTex = this._device.createTexture(new TextureInfo(
                TextureType.TEX2D,
                TextureUsageBit.STORAGE | TextureUsageBit.TRANSFER_SRC | TextureUsageBit.SAMPLED,
                Format.RGBA8,
                texWidth,
                texHeight,
            ));

            this.descriptorSet.bindTexture(UNIFORM_REFLECTION_TEXTURE_BINDING, this._reflectionTex);

            this._reflectionSampler = this._device.getSampler(new SamplerInfo(
                Filter.LINEAR,
                Filter.LINEAR,
                Filter.NONE,
                Address.CLAMP,
                Address.CLAMP,
                Address.CLAMP,
            ));
            this.descriptorSet.bindSampler(UNIFORM_REFLECTION_TEXTURE_BINDING, this._reflectionSampler);
            this.descriptorSet.bindTexture(UNIFORM_REFLECTION_STORAGE_BINDING, this._reflectionTex);
        }
    }

    /**
     * @en
     * destroy sub model
     * @zh
     * 销毁子模型
     */
    public destroy (): void {
        this._descriptorSet!.destroy();
        this._descriptorSet = null;

        this._inputAssembler!.destroy();
        this._inputAssembler = null;

        this._worldBoundDescriptorSet?.destroy();
        this._worldBoundDescriptorSet = null;

        this.priority = RenderPriority.DEFAULT;

        this._patches = null;
        this._subMesh = null;

        this._passes = null;
        this._shaders = null;

        if (this._reflectionTex) this._reflectionTex.destroy();
        this._reflectionTex = null;
        this._reflectionSampler = null;
    }

    /**
     * @en
     * update sub model
     * @zh
     * 更新子模型
     */
    public update (): void {
        for (let i = 0; i < this._passes!.length; ++i) {
            const pass = this._passes![i];
            pass.update();
        }
        this._descriptorSet!.update();
        this._worldBoundDescriptorSet?.update();
    }

    /**
     * @en Pipeline changed callback
     * @zh 管线更新回调
     */
    public onPipelineStateChanged (): void {

        const passes = this._passes;
        if (!passes) { return; }

        for (let i = 0; i < passes.length; i++) {
            const pass = passes[i];
            pass.beginChangeStatesSilently();
            pass.tryCompile(); // force update shaders
            pass.endChangeStatesSilently();
        }

        this._flushPassInfo();
    }

    /**
     * @en Shader macro changed callback
     * @zh Shader 宏更新回调
     */
    public onMacroPatchesStateChanged (patches: IMacroPatch[] | null): void {
        if (!patches && !this._patches) {
            return;
        } else if (patches) {
            patches = patches.sort();
            // Sorting on shorter patches outperforms hashing, with negative optimization on longer global patches.
            if (this._patches && patches.length === this._patches.length) {
                const patchesStateUnchanged = JSON.stringify(patches) === JSON.stringify(this._patches);
                if (patchesStateUnchanged) return;
            }
        }
        this._patches = patches;

        const passes = this._passes;
        if (!passes) { return; }

        for (let i = 0; i < passes.length; i++) {
            const pass = passes[i];
            pass.beginChangeStatesSilently();
            pass.tryCompile(); // force update shaders
            pass.endChangeStatesSilently();
        }

        this._flushPassInfo();
    }

    /**
     * @en
     * geometry changed callback
     * @zh
     * 几何数据改变后的回调
     */
    public onGeometryChanged (): void {
        if (!this._subMesh) {
            return;
        }

        // update draw info
        const drawInfo = this._subMesh.drawInfo;

        // to invoke getter/setter function for wasm object
        if (this._inputAssembler && drawInfo) {
            const dirtyDrawInfo = this._inputAssembler.drawInfo;
            Object.keys(drawInfo).forEach((key): void => {
                dirtyDrawInfo[key] = drawInfo[key];
            });
            this._inputAssembler.drawInfo = dirtyDrawInfo;
        }
    }

    /**
     * @en
     * get instanced attribute index
     * @zh
     * 获取硬件实例化相关索引
     */
    /**
     * @internal
     */
    public getInstancedAttributeIndex (name: string): number {
        const { attributes } = this.instancedAttributeBlock;
        for (let i = 0; i < attributes.length; i++) {
            if (attributes[i].name === name) { return i; }
        }
        return -1;
    }

    /**
     * @en
     * update instancing related data, invoked by model
     * @zh
     * 更新硬件实例化相关数据，一般由model调用
     */
    /**
     * @internal
     */
    public updateInstancedWorldMatrix (mat: Mat4, idx: number): void {
        const attrs = this.instancedAttributeBlock.views;
        const v1 = attrs[idx];
        const v2 = attrs[idx + 1];
        const v3 = attrs[idx + 2];
        v1[0] = mat.m00; v1[1] = mat.m01; v1[2] = mat.m02; v1[3] = mat.m12;
        v2[0] = mat.m04; v2[1] = mat.m05; v2[2] = mat.m06; v2[3] = mat.m13;
        v3[0] = mat.m08; v3[1] = mat.m09; v3[2] = mat.m10; v3[3] = mat.m14;
    }

    /**
     * @en
     * update instancing SH data, invoked by model
     * @zh
     * 更新硬件实例化球谐数据，一般由model调用
     */
    /**
     * @internal
     */
    public updateInstancedSH (data: Float32Array, idx: number): void {
        const attrs = this.instancedAttributeBlock.views;
        const count = (UBOSH.SH_QUADRATIC_R_OFFSET - UBOSH.SH_LINEAR_CONST_R_OFFSET) / 4;
        let offset = 0;

        for (let i = idx; i < idx + count; i++) {
            for (let k = 0; k < 4; k++) {
                attrs[i][k] = data[offset++];
            }
        }
    }

    /**
     * @en
     * update instancing related data, invoked by model
     * @zh
     * 更新硬件实例化相关数据，一般由model调用
     */
    /**
     * @internal
     */
    public UpdateInstancedAttributes (attributes: Attribute[]): void {
        // initialize subModelWorldMatrixIndex
        this.instancedWorldMatrixIndex = -1;
        this.instancedSHIndex = -1;

        const pass = this.passes[0];
        if (!pass.device.hasFeature(Feature.INSTANCED_ARRAYS)) { return; }
        // free old data

        let size = 0;
        for (let j = 0; j < attributes.length; j++) {
            const attribute = attributes[j];
            if (!attribute.isInstanced) { continue; }
            size += FormatInfos[attribute.format].size;
        }

        const attrs = this.instancedAttributeBlock;
        attrs.buffer = new Uint8Array(size);
        attrs.views.length = attrs.attributes.length = 0;
        let offset = 0;
        for (let j = 0; j < attributes.length; j++) {
            const attribute = attributes[j];
            if (!attribute.isInstanced) { continue; }
            const attr = new Attribute();
            attr.format = attribute.format;
            attr.name = attribute.name;
            attr.isNormalized = attribute.isNormalized;
            attr.location = attribute.location;
            attrs.attributes.push(attr);

            const info = FormatInfos[attribute.format];

            const typeViewArray = new (getTypedArrayConstructor(info))(attrs.buffer.buffer, offset, info.count);
            attrs.views.push(typeViewArray);
            offset += info.size;
        }
        if (pass.batchingScheme === BatchingSchemes.INSTANCING) { pass.getInstancedBuffer().destroy(); } // instancing IA changed
        this.instancedWorldMatrixIndex = this.getInstancedAttributeIndex(INST_MAT_WORLD);
        this.instancedSHIndex = this.getInstancedAttributeIndex(INST_SH);
    }

    protected _flushPassInfo (): void {
        const passes = this._passes;
        if (!passes) { return; }
        if (!this._shaders) { this._shaders = []; }

        this._shaders.length = passes.length;
        for (let i = 0, len = passes.length; i < len; i++) {
            this._shaders[i] = passes[i].getShaderVariant(this.patches)!;
        }
    }
}
